R/helper_patient_data_fix.R

```diff
@@ -172,27 +172,89 @@ parse_dates <- function(date) {
 
 #' @title Check value against list of allowed values.
 #'
+#' @description
+#' Invalid values are logged and depending on the replace invalid flag
+#' also replaced with an error value. When checking for the validity of values
+#' the case of the string is not considered nor are spaces and non-alphanumeric
+#' characters. If the value matches an entry in the allowed_values list
+#' that match is returned as the result.
+#'
 #' @param x character value to check.
 #' @param valid_values list of valid options.
-#' @param error_val value that is used to signal invalid values.
 #' @param id patient id.
+#' @param replace_invalid A flag controlling whether to also replace invalid .
+#' @param error_val value that is used to signal invalid values.
 #' @param col column name when used with mutate/across.
 #'
 #' @return Either x or "Other".
-check_allowed_values <- function(x, valid_values, error_val, id, col = "") {
+check_allowed_values <- function(x, valid_values, id, replace_invalid = TRUE, error_val = "Undefined", col = "") {
     if (is.na(x) || x == "") {
         return(NA_character_)
     }
 
-    if (!tolower(x) %in% tolower(valid_values)) {
+    valid_value_mapping <- setNames(as.list(valid_values), sanitize_str(valid_values))
+
+    if (!sanitize_str(x) %in% names(valid_value_mapping)) {
         logWarn("Patient ", id, ": Value ", x, " for column ", col, " is not in the list of allowed values. ")
-        if (!is.na(error_val)) {
+        if (replace_invalid) {
             logInfo("Replacing ", x, " with ", error_val, ".")
-            x <- error_val
+            return(error_val)
+        } else {
+            return(x)
         }
     }
 
-    x
+    valid_value_mapping[[sanitize_str(x)]]
+}
+
+parse_allowed_value_check <- function(column_name, check_details) {
+    if ("error_value" %in% check_details) {
+        error_expr <- rlang::parse_expr(check_details$error_value)
+        check <- rlang::expr(
+            check_allowed_values(
+                !!check_details$allowed_values,
+                id,
+                !!check_details$replace_invalid,
+                !!error_expr,
+                !!column_name
+            )
+        )
+    } else {
```

Comment:
this is a lot of code repetition. Maybe we can just use `if else` to set the value for the error_expr, and call the function only once because nothing else changes?